iree/hal/vulkan/native_timeline_semaphore.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/vulkan/native_timeline_semaphore.h"
+
+#include "iree/base/tracing.h"
+#include "iree/hal/vulkan/status_util.h"
+
+namespace iree {
+namespace hal {
+namespace vulkan {
+
+NativeTimelineSemaphore::NativeTimelineSemaphore(
+    ref_ptr<VkDeviceHandle> logical_device, VkSemaphore handle,
+    uint64_t initial_value)
+    : logical_device_(std::move(logical_device)), handle_(handle) {
+  IREE_TRACE_SCOPE0("NativeTimelineSemaphore::ctor");
+}
+
+NativeTimelineSemaphore::~NativeTimelineSemaphore() {
+  IREE_TRACE_SCOPE0("NativeTimelineSemaphore::dtor");
+  logical_device_->syms()->vkDestroySemaphore(*logical_device_, handle_,
+                                              logical_device_->allocator());
+}
+
+StatusOr<uint64_t> NativeTimelineSemaphore::Query() {
+  uint64_t value = 0;
+  VK_RETURN_IF_ERROR(logical_device_->syms()->vkGetSemaphoreCounterValue(
+      *logical_device_, handle_, &value));
+  if (value == UINT64_MAX) {
+    absl::MutexLock lock(&status_mutex_);
+    return status_;
+  }
+  return value;
+}
+
+Status NativeTimelineSemaphore::Signal(uint64_t value) {
+  IREE_TRACE_SCOPE0("NativeTimelineSemaphore::Signal");
+
+  VkSemaphoreSignalInfo signal_info;
+  signal_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_SIGNAL_INFO;
+  signal_info.pNext = nullptr;
+  signal_info.semaphore = handle_;
+  signal_info.value = value;
+  return VkResultToStatus(logical_device_->syms()->vkSignalSemaphore(
+      *logical_device_, &signal_info));
+}
+
+void NativeTimelineSemaphore::Fail(Status status) {
+  IREE_TRACE_SCOPE0("NativeTimelineSemaphore::Fail");
+
+  // NOTE: we hold the lock here as the vkSignalSemaphore may wake a waiter and
+  // we want to be able to immediately give them the status.
+  absl::MutexLock lock(&status_mutex_);
+  status_ = std::move(status);
+
+  VkSemaphoreSignalInfo signal_info;
+  signal_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_SIGNAL_INFO;
+  signal_info.pNext = nullptr;
+  signal_info.semaphore = handle_;
+  signal_info.value = UINT64_MAX;
+  // NOTE: we don't care about the result in case of failures as we are
+  // failing and the caller will likely be tearing everything down anyway.
+  logical_device_->syms()->vkSignalSemaphore(*logical_device_, &signal_info);
+}
+
+Status NativeTimelineSemaphore::Wait(uint64_t value, absl::Time deadline) {
```

Comment:
I see that some of this logic is duplicated in `VulkanDevice::WaitSemaphores`. If that logic was moved into `NativeTimelineSemaphore` here, then an `EmulatedTimelineSemaphore` (not using the extension layer) could be slotted in more easily. WDYT?

Reply:
Single-wait and multi-wait will sometimes have different implementations (though don't in vulkan), but usually the multi-wait is the load-bearing one. The important thing there is that there may be multiple kinds of semaphores passed in: for example, a host semaphore + a device semaphore + imported fd semaphore all being waited on together. I think the real fix here for when we have that heterogeneous semaphore situation is using WaitHandle (or really a derivative of it, as it's pretty old at this point), which was designed to allow this mixed case.

For now I'll just make these single-wait methods call the multi-wait to reduce duplication, but it'll require some special casing for other types (like until we have a mixed use mode just call static NativeTimelineSemaphore::MultiWait, etc). Can do that when we get an emulated timeline semaphore :)

Reply:
Nevermind, I forgot the layering: this intentionally doesn't have a Device handle so it can't call the multi-wait method. Will just leave as-is for now.